pkg/cache/cluster.go

```diff
@@ -779,25 +969,35 @@ func (c *clusterCache) onNodeUpdated(oldRes *Resource, un *unstructured.Unstruct
 	newRes := c.newResource(un)
 	c.setNode(newRes)
 	for _, h := range c.getResourceUpdatedHandlers() {
-		h(newRes, oldRes, c.nsIndex[newRes.Ref.Namespace])
+		r, _ := c.LoadNsIndex(newRes.Ref.Namespace)
+		h(newRes, oldRes, r)
 	}
 }
 
-func (c *clusterCache) onNodeRemoved(key kube.ResourceKey) {
-	existing, ok := c.resources[key]
-	if ok {
-		delete(c.resources, key)
-		ns, ok := c.nsIndex[key.Namespace]
-		if ok {
-			delete(ns, key)
-			if len(ns) == 0 {
-				delete(c.nsIndex, key.Namespace)
+func (c *clusterCache) onNodeRemoved(key kube.ResourceKey) error {
+	existing, err := c.resources.LoadResources(key)
```

Comment:
I agree with @ash2k regarding `sync.Map` usage.

I think we still need to have a lock here to make sure `resources` map and namespace index are synchronized.